### PR TITLE
feat: highlight markdown code blocks

### PR DESCRIPTION
**Pull request by magicmatatjahu**
**Description**

- highlight markdown code blocks using `highlight.js`:
  - load dynamically config for languages used in given specification
  - write helpers for it
  - write tests for it
- use latest `react-component` with standalone bundle:
  - remove from script `react` and `react-dom` packages
  
Examples:

C#:
image

elixir
image

<details>
  <summary>Example spec for testing:</summary>

```yaml
asyncapi: '2.0.0'
id: 'urn:com:smartylighting:streetlights:server'
info:
  title: Streetlights API
  version: '1.0.0'
  description: |
    The Smartylighting Streetlights API allows you to remotely manage the city lights.

    ### Check out its awesome features:

    * Turn a specific streetlight on/off 🌃
    * Dim a specific streetlight 😎
    * Receive real-time information about environmental lighting conditions 📈

  termsOfService: http://asyncapi.org/terms/
  contact:
    name: API Support
    url: http://www.asyncapi.org/support
    email: support@asyncapi.org
  license:
    name: Apache 2.0
    url: http://www.apache.org/licenses/LICENSE-2.0.html
tags:
  - name: root-tag1
    externalDocs:
      description: External docs description 1
      url: https://www.asyncapi.com/
  - name: root-tag2
    description: Description 2
    externalDocs:
      url: "https://www.asyncapi.com/"
  - name: root-tag3
  - name: root-tag4
    description: Description 4
  - name: root-tag5
    externalDocs:
      url: "https://www.asyncapi.com/"
externalDocs:
  description: Find more info here
  url: https://example.com
defaultContentType: application/json

servers:
  production:
    url: api.streetlights.smartylighting.com:{port}
    protocol: mqtt
    description: |
      Private server that requires authorization.
      Once the socket is open you can subscribe to private-data channels by sending an authenticated subscribe request message.

      The API client must request an authentication "token" via the following REST API endpoint "GetWebSocketsToken" to connect to WebSockets Private endpoints. For more details read https://support.kraken.com/hc/en-us/articles/360034437672-How-to-retrieve-a-WebSocket-authentication-token-Example-code-in-Python-3

      The resulting token must be provided in the "token" field of any new private WebSocket feed subscription: 
      ```json
      {
        "event": "subscribe",
        "subscription":
        {
          "name": "ownTrades",
          "token": "WW91ciBhdXRoZW50aWNhdGlvbiB0b2tlbiBnb2VzIGhlcmUu"
        }
      }
      ```

      ```elixir
      defmodule Hello do
        def world do
          IO.puts("hello")
        end
      end
      ```
    variables:
      port:
        description: Secure connection (TLS) is available through port 8883.
        default: '1883'
        enum:
          - '1883'
          - '8883'
    security:
      - apiKey: []
      - supportedOauthFlows:
        - streetlights:on
        - streetlights:off
        - streetlights:dim
      - openIdConnectWellKnown: []
  dummy-mqtt:
    url: mqtt://localhost
    protocol: mqtt
    description: |
      Private server

      ```csharp
      using System;

      namespace HelloWorld
      {
        class Program
        {
          static void Main(string[] args)
          {
            Console.WriteLine("Hello World!");    
          }
        }
      }
      ```
    bindings:
      mqtt:
        clientId: guest        
        cleanSession: false
        keepAlive: 60
        bindingVersion: 0.1.0
        lastWill:
          topic: smartylighting/streetlights/1/0/lastwill
          qos: 1
          message: so long and thanks for all the fish
          retain: false
  dummy-amqp:
    url: amqp://localhost:{port}
    protocol: amqp
    description: dummy AMQP broker
    protocolVersion: "0.9.1"
    variables:
      port:
        enum:
          - '15672'
          - '5672'
  dommy-kafka:
    url: http://localhost:{port}
    protocol: kafka
    description: dummy Kafka broker
    variables:
      port:
        default: '9092'

channels:
  smartylighting/streetlights/1/0/event/{streetlightId}/lighting/measured:
    description: The topic on which measured values may be produced and consumed.
    parameters:
      streetlightId:
        $ref: '#/components/parameters/streetlightId'
    subscribe:
      summary: Receive information about environmental lighting conditions of a particular streetlight.
      operationId: receiveLightMeasurement
      externalDocs:
        description: Find more info here
        url: https://example.com
      traits:
        - $ref: '#/components/operationTraits/kafka'
      message:
        $ref: '#/components/messages/lightMeasured'
      bindings:
        mqtt:
          qos: 1
          bindingVersion: 0.1.0
        http:
          type: request
          method: GET
          query:
            type: object
            required:
            - companyId
            properties:
              companyId:
                type: number
                minimum: 1
                description: The Id of the company.
            additionalProperties: false

  smartylighting/streetlights/1/0/action/{streetlightId}/turn/on:
    parameters:
      streetlightId:
        $ref: '#/components/parameters/streetlightId'
    publish:
      operationId: turnOn
      externalDocs:
        description: Find more info here
        url: https://example.com
      traits:
        - $ref: '#/components/operationTraits/kafka'
      message:
        $ref: '#/components/messages/turnOnOff'

  smartylighting/streetlights/1/0/action/{streetlightId}/turn/off:
    parameters:
      streetlightId:
        $ref: '#/components/parameters/streetlightId'
    publish:
      traits:
        - $ref: '#/components/operationTraits/kafka'
      message:
        $ref: '#/components/messages/turnOnOff'

  smartylighting/streetlights/1/0/action/{streetlightId}/dim:
    parameters:
      streetlightId:
        $ref: '#/components/parameters/streetlightId'
    publish:
      operationId: dimLight
      traits:
        - $ref: '#/components/operationTraits/kafka'
      message:
        $ref: '#/components/messages/dimLight'

components:
  messages:
    lightMeasured:
      name: lightMeasured
      title: Light measured
      summary: Inform about environmental lighting conditions for a particular streetlight.
      contentType: application/json
      correlationId:
        $ref: "#/components/correlationIds/sentAtCorrelator"
      externalDocs:
        url: "https://www.asyncapi.com/"
      traits:
        - $ref: '#/components/messageTraits/commonHeaders'
      payload:
        $ref: "#/components/schemas/lightMeasuredPayload"
      bindings:
        mqtt:
          bindingVersion: 0.1.0
      examples:
        - headers:
            my-app-header: 12
          payload:
            lumens: 1
            sentAt: "2020-01-31T13:24:53Z"
        - headers:
            my-app-header: 13
        - payload:
            lumens: 3
            sentAt: "2020-10-31T13:24:53Z"
      x-schema-extensions-as-object:
        type: object
        properties:
          prop1:
            type: string
          prop2:
            type: integer
            minimum: 0
      x-schema-extensions-as-primitive: dummy
      x-schema-extensions-as-array: 
        - "item1"
        - "item2"
    LwM2mOjbects:
      payload:
        type: object
        properties:
          objectLinks:
            type: string
        example:
          objectLinks: "lwm2m=1.1, </0/0>, </1/1>;ssid=1, </2>, </3/0>"
    turnOnOff:
      name: turnOnOff
      title: Turn on/off
      summary: Command a particular streetlight to turn the lights on or off.
      payload:
        $ref: "#/components/schemas/turnOnOffPayload"
      headers: 
        type: object
        properties:
          $ref: '#/components/schemas/streamHeaders'
    dimLight:
      name: dimLight
      title: Dim light
      summary: Command a particular streetlight to dim the lights.
      correlationId:
        $ref: "#/components/correlationIds/sentAtCorrelator"
      externalDocs:
        url: "https://www.asyncapi.com/"
      tags:
        - name: oparation-tag1
          externalDocs:
            description: External docs description 1
            url: https://www.asyncapi.com/
        - name: oparation-tag2
          description: Description 2
          externalDocs:
            url: "https://www.asyncapi.com/"
        - name: oparation-tag3
        - name: oparation-tag4
          description: Description 4
        - name: oparation-tag5
          externalDocs:
            url: "https://www.asyncapi.com/"
      traits:
        - $ref: '#/components/messageTraits/commonHeaders'
      payload:
        $ref: "#/components/schemas/dimLightPayload"

  schemas:
    lightMeasuredPayload:
      type: object
      properties:
        lumens:
          type: integer
          description: Light intensity measured in lumens.
          writeOnly: true
          oneOf: 
            - minimum: 0
              maximum: 5
            - minimum: 10
              maximum: 20
          externalDocs:
            url: "https://www.asyncapi.com/"
        sentAt:
          $ref: "#/components/schemas/sentAt"
        ifElseThen:
          type: integer
          minimum: 1
          maximum: 1000
          if:
            minimum: 100
          then: 
            multipleOf: 100
          else:
            if: 
              minimum: 10
            then: 
              multipleOf: 10
              description: |
                ```json
                {
                  "event": "subscribe",
                  "subscription":
                  {
                    "name": "ownTrades",
                    "token": "WW91ciBhdXRoZW50aWNhdGlvbiB0b2tlbiBnb2VzIGhlcmUu"
                  }
                }
                ```

                ```elixir
                defmodule Hello do
                  def world do
                    IO.puts("hello")
                  end
                end
                ```
        dependencies:
          $ref: "#/components/schemas/dependenciesObject"
      required:
        - lumens
      x-schema-extensions-as-object:
        type: object
        properties:
          prop1:
            type: string
          prop2:
            type: integer
            minimum: 0
      x-schema-extensions-as-primitive: dummy
      x-schema-extensions-as-array: 
        - "item1"
        - "item2"
    turnOnOffPayload:
      type: object
      properties:
        command:
          type: string
          enum:
            - on
            - off
          description: Whether to turn on or off the light.
        sentAt:
          $ref: "#/components/schemas/sentAt"
      additionalProperties:
        type: string

    dimLightPayload:
      type: object
      properties:
        percentage:
          type: integer
          description: Percentage to which the light should be dimmed to.
          minimum: 0
          maximum: 100
          readOnly: true
        sentAt:
          $ref: "#/components/schemas/sentAt"
        key:
          type: integer
          not:
            minimum: 3
      patternProperties:
        ^S_:
          type: string
        ^I_:
          type: integer
      additionalProperties: false
    sentAt:
      type: string
      format: date-time
      description: Date and time when the message was sent.
    union:
      type: [string, number]
    objectWithKey:
      type: object
      propertyNames:
        format: email
      properties:
        key:
          type: string
    objectWithKey2:
      type: object
      properties:
        key2:
          type: string
          format: time
    oneOfSchema:
      oneOf:
        - $ref: "#/components/schemas/objectWithKey"
        - $ref: "#/components/schemas/objectWithKey2"
    anyOfSchema:
      anyOf:
        - $ref: "#/components/schemas/objectWithKey"
        - $ref: "#/components/schemas/objectWithKey2"
    allOfSchema:
      allOf:
        - $ref: "#/components/schemas/objectWithKey"
        - $ref: "#/components/schemas/objectWithKey2"
    arrayContains: 
      type: array
      contains:
        type: integer
    dependenciesObject:
      type: object
      properties:
        name:
          type: string
        credit_card:
          type: integer
        billing_address:
          type: string
        schema_dependency:
          type: string
      required:
        - name
      dependencies:
        credit_card:
          properties:
            billing_address:
              type: string
            billing_address2:
              type: string
          required:
          - billing_address
          dependencies:
            billing_address2:
              properties:
                billing_address3:
                  type: string
              required:
              - billing_address3    

    subscriptionStatus:
      type: object
      oneOf:
        - properties:
            channelID:
              type: integer
              description: ChannelID on successful subscription, applicable to public messages only.
            channelName:
              type: string
              description: Channel Name on successful subscription. For payloads 'ohlc' and 'book', respective interval or depth will be added as suffix.
        - properties:
            errorMessage:
              type: string
      properties:
        event:
          type: string
          const: subscriptionStatus
        subscription:
          type: object
          properties:
            depth:
              type: string
            interval:
              type: string
          required:
            - name
      required:
        - event

    streamHeaders:
      Etag:
        type: string
        description: |
          The RFC7232 ETag header field in a response provides the current entity-
          tag for the selected resource. An entity-tag is an opaque identifier for
          different versions of a resource over time, regardless whether multiple
          versions are valid at the same time. An entity-tag consists of an opaque
          quoted string, possibly prefixed by a weakness indicator.
        example: 411a
      Cache-Control:
        description: The Cache-Control HTTP header holds directives (instructions) for caching in request.
        type: string
        example: no-cache, no-store, must-revalidate

  securitySchemes:
    apiKey:
      type: apiKey
      in: user
      description: Provide your API key as the user and leave the password empty.
    supportedOauthFlows:
      type: oauth2
      description: Flows to support OAuth 2.0
      flows:
        implicit:
          authorizationUrl: 'https://authserver.example/auth'
          scopes:
            'streetlights:on': Ability to switch lights on
            'streetlights:off': Ability to switch lights off
            'streetlights:dim': Ability to dim the lights
        password:
          tokenUrl: 'https://authserver.example/token'
          scopes:
            'streetlights:on': Ability to switch lights on
            'streetlights:off': Ability to switch lights off
            'streetlights:dim': Ability to dim the lights
        clientCredentials:
          tokenUrl: 'https://authserver.example/token'
          scopes:
            'streetlights:on': Ability to switch lights on
            'streetlights:off': Ability to switch lights off
            'streetlights:dim': Ability to dim the lights
        authorizationCode:
          authorizationUrl: 'https://authserver.example/auth'
          tokenUrl: 'https://authserver.example/token'
          refreshUrl: 'https://authserver.example/refresh'
          scopes:
            'streetlights:on': Ability to switch lights on
            'streetlights:off': Ability to switch lights off
            'streetlights:dim': Ability to dim the lights
    openIdConnectWellKnown:
      type: openIdConnect
      openIdConnectUrl: 'https://authserver.example/.well-known'

  parameters:
    streetlightId:
      description: The ID of the streetlight.
      schema:
        type: string
      location: "$message.payload#/user/id"

  correlationIds:
    sentAtCorrelator:
      description: Data from message payload used as correlation ID
      location: $message.payload#/sentAt

  messageTraits:
    commonHeaders:
      headers:
        type: object
        properties:
          my-app-header:
            type: integer
            minimum: 0
            maximum: 100
        required:
          - my-app-header
  
  operationTraits:
    kafka:
      bindings:
        kafka:
          clientId: my-app-id
```

</details>

**Related issue(s)**
Part of https://github.com/asyncapi/asyncapi-react/issues/354
Blocked by https://github.com/asyncapi/asyncapi-react/pull/361

### CODE REVIEW

**Comment by jonaslagoni** [comment]
@magicmatatjahu you have a failing test 🙂

**Comment by magicmatatjahu** [comment]
@jonaslagoni The first one in `react-component` isn't merged yet, so we need to wait for it :)

**Comment by magicmatatjahu** [comment]
@jonaslagoni `react-component` is updated to latest version :)

File: test/filters/all.test.js [hunk 1]
```diff
@@ -79,11 +141,11 @@ describe('Filters', () => {
...
     it('should work', async () => {
-      const doc = new AsyncAPIDocument({ asyncapi: '2.0.0', info: { title: 'dummy spec for testing', version: '1.5.73' } });
+      const doc = new AsyncAPIDocument({ asyncapi: '2.0.0', info: { title: 'dummy spec for testing', version: '2.1.37' } });
 
       const result = renderSpec(doc);
-      // check if '1.5.73' version is rendered
-      expect(result.includes('1.5.73')).toEqual(true);
+      // check if '2.1.37' version is rendered
+      expect(result.includes('2.1.37')).toEqual(true);
       // check if 'dummy spec for testing' title is rendered
```
Comment (jonaslagoni) [inline test/filters/all.test.js:149]: Why was this changed? 🤔

File: test/filters/all.test.js [hunk 2]
```diff
@@ -1,11 +1,73 @@
...
+      expect(hljs.getLanguage('ts') === undefined).toEqual(false);
+      expect(hljs.getLanguage('typescript') === hljs.getLanguage('ts')).toEqual(true);
+    });
+
+    it('should warn if at least one language does not exist', async () => {
+      const originalWarn = console.warn;
+      let warning;
+      console.warn = (...args) => { warning = args[0] };
+      loadLanguagesConfig('text\n```bash\nfoo\n```\ntext abc\n```foo_bar\nbar\n```\nxyz');
+      console.warn = originalWarn;
```
Comment (jonaslagoni) [inline test/filters/all.test.js:65]: You could mock the function here if you wanted, easier to restore the previous state as well.

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Added a few comments, looks good overall 👍

File: test/filters/all.test.js (hunk 1 above)
Comment (magicmatatjahu) [reply test/filters/all.test.js:149]: Do you not like the `2.1.37`? 😄 It's only a test data.

File: filters/all.js [hunk 3]
```diff
@@ -76,6 +91,49 @@ function includeFile(pathFile) {
 }
 filter.includeFile = includeFile;
 
+/**
+ * Retrieves all languages included in code blocks in the specification.
+ */
+function retrieveLanguages(originalAsyncAPI) {
+  const regex = /^(([ \t]*`{3,4})([^\n]*)([\s\S]+?)(^[ \t]*\2))/gm,
```
[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply filters/all.js:106]: Added :) Thanks!

File: test/filters/all.test.js [hunk 4]
```diff
@@ -79,11 +148,11 @@ describe('Filters', () => {
...
   describe('.renderSpec', () => {
     it('should work', async () => {
-      const doc = new AsyncAPIDocument({ asyncapi: '2.0.0', info: { title: 'dummy spec for testing', version: '1.5.73' } });
+      const doc = new AsyncAPIDocument({ asyncapi: '2.0.0', info: { title: 'dummy spec for testing', version: '1.5.34' } });
 
       const result = renderSpec(doc);
-      // check if '1.5.73' version is rendered
-      expect(result.includes('1.5.73')).toEqual(true);
+      // check if '1.5.34' version is rendered
+      expect(result.includes('1.5.34')).toEqual(true);
```
Comment (jonaslagoni) [inline test/filters/all.test.js:155]: Are you mocking me or 😂? I know it is test data, but when there is no need to change it, it just clutter the changes 😄

**Review by jonaslagoni — APPROVED** [review]
LGTM 👍

**Review by derberg — COMMENTED** [review]
why do we need such changes here in `html-template`, why it is not part of the react component?

**Comment by magicmatatjahu** [comment]
@derberg In React component we have only logic related to render. Also we don't have included all configuration for langauges, because they weight about 800-900 kbs, but only three configurations for `json`, `yaml` and `bash` to reduce the size of component - why? More info is in https://github.com/asyncapi/asyncapi-react/pull/361 In this PR I added logic to retrieve all languages from spec and load their configuration in rendering phase.

**Comment by derberg** [comment]
makes sense, so we have highlighting in react component but with limited scope, sounds fair. It just needs to be clearly noted in description of the PR + readme of react component as it is not that straightforward and we will have confusion in future, that in Playground folks will see full highlight support and then limited one when they use react component directly. Then we move to studio where we use react component directly and we either forget or add this extended functionality there - thus, for both still confusing

**Comment by magicmatatjahu** [comment]
I added docs for it in the `next` branch :) https://github.com/asyncapi/asyncapi-react/blob/next/docs/features/highlight.md For studio as it will be pure JS app running in browser we have several options:
- load only common languages like java, ruby, c# etc, highlight.js has common plugin to add config for these languages, there are 20+
- add all languages but not directly in the final main bundle but make chunks and load these chunks only when someone would render the spec with given very rare language
Only as you wrote, we need to know and remember about that.

**Comment by magicmatatjahu** [comment]
@jonaslagoni I had to change the way of loading the configuration. At the moment we load everything. It's faster and maybe it's surprise, we don't increase the size of the rendered file because the configuration works only in the rendering phase and then nothing is added :)

File: generate.yaml [hunk 5]
```diff
@@ -0,0 +1,567 @@
+asyncapi: '2.0.0'
```
Comment (jonaslagoni) [inline generate.yaml:1]: Was this file supposed to be committed? It has no references.

**Review by jonaslagoni — COMMENTED** [review]
Just had 1 comment 🙂

File: generate.yaml (hunk 5 above)
Comment (magicmatatjahu) [reply generate.yaml]: I used it for testing and forgot about it. Removed, thanks!

**Review by jonaslagoni — APPROVED** [review]
👍

**Comment by magicmatatjahu** [comment]
@jonaslagoni Thanks!
